### PR TITLE
Transform codebase to typescript

### PR DESCRIPTION
**Pull request by radek-stary**

### CODE REVIEW

**Review by mbohal — CHANGES_REQUESTED** [review]

File: webpack.config.js [hunk 1]
```diff
@@ -69,7 +74,7 @@ module.exports = (env, argv) => ({
       ? '[name].js'
       : '[name].development.js',
     libraryTarget: 'umd',
-    path: path.join(__dirname, 'dist'),
+     path: path.join(__dirname, 'dist'),
```
[the remark this answers is not in the document]
Comment (adamkudrna) [reply webpack.config.js:77]: Why did ESlint miss that?

File: tests/providerTests/formLayoutProviderTest.tsx
[the remark this answers is not in the document]
Comment (adamkudrna) [reply tests/providerTests/formLayoutProviderTest.tsx:1]: @mbohal Both your commands are identical 🙂.

File: .eslintrc [hunk 2]
```diff
@@ -14,9 +33,5 @@
...
       ]
     }
-  ],
-  "parser": "@babel/eslint-parser",
-  "rules": {
-    "import/prefer-default-export": "off"
-  }
-}
+  ]
+}
```
Comment (adamkudrna) [inline .eslintrc:37]: Please fix the new line at the end of the file.

File: definitions.d.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/**
+ * [Color variant](/foundation/colors#component-colors)
+ */
+declare type Color = 'primary' | 'secondary' | 'selected' | 'success' | 'warning' | 'danger' | 'help' | 'info' | 'note' | 'light' | 'dark';
```
Comment (adamkudrna) [inline definitions.d.ts:4]: ⚠️ Not all colors can be always used. There are multiple "dictionaries" that group colors suitable for certain use case, see [RUI docs](https://react-ui.io/foundation/colors#applying-colors).

Does it make sense to separate them in TS? (I think it does, but maybe not now.)

File: definitions.d.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+declare type Placement = 'top' | 'top-start' | 'top-end' | 'right' | 'right-start' | 'right-end' | 'bottom' | 'bottom-start' | 'bottom-end' | 'left' | 'left-start' | 'left-end';
+declare type AutoScroll = 'always' | 'detectEnd' | 'off';
+declare type Direction = 'asc' | 'desc';
+declare type Hyphens = 'none' | 'auto' | 'manual';
+declare type WordWrapping = 'normal' | 'long-words' | 'anywhere';
+declare type VerticalModalPosition = 'top' | 'center';
+
+declare type PredefinedLabelWidthValues = 'auto' | 'default' | 'limited';
+
+// Test types
```
Comment (adamkudrna) [inline definitions.d.ts:40]: Why a different type of comment? I mean `//` vs. `/* */`.

Secondly, why only some types have comments and some don't?

File: jest.config.js [hunk 5]
```diff
@@ -2,19 +2,25 @@ module.exports = {
   moduleFileExtensions: [
     'js',
```
[the remark this answers is not in the document]
Comment (adamkudrna) [reply jest.config.js:3]: If changing here, please also adjust `transform` below.

File: package.json [hunk 6]
```diff
@@ -33,7 +33,8 @@
     "npm": ">=8.3.0"
   },
   "dependencies": {
-    "normalize.css": "^8.0.1"
+    "normalize.css": "^8.0.1",
+    "ts-jest": "^27.1.5"
```
Comment (adamkudrna) [inline package.json:37]: Isn't it a dev dependency?

File: src/docs/_components/Icon/Icon.types.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+export interface IconProps {
+  icon: 'code' |
+  'down' |
+  'loading' |
+  'pencil' |
+  'star' |
+  'success' |
+  'up' |
+  'warning-sign'
```
Comment (adamkudrna) [inline src/docs/_components/Icon/Icon.types.ts:9]: Shouldn't these be intended?

File: src/docs/_components/Placeholder/Placeholder.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import { classNames } from '../../../lib/utils/classNames';
+import styles from './Placeholder.scss';
+import {
+  PlaceholderProps, CustomCSSProperties,
+} from './Placeholder.types';
```
Comment (adamkudrna) [inline src/docs/_components/Placeholder/Placeholder.tsx:6]: I think this should either be a single line, or each item should have its own line. Why ESlint doesn't cover it…? @bedrich-schindler @mbohal

File: src/lib/components/Alert/Alert.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        styles.root,
+        getRootColorClassName(color, styles),
+      )}
+      id={id}
+      role="alert"
+    >
+      {icon && (
+      <div className={styles.icon}>
+        {icon}
+      </div>
```
Comment (adamkudrna) [inline src/lib/components/Alert/Alert.tsx:35]: These lines should be indented.

File: tests/propTests/sizePropTest.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  ],
+  [
+    { size: 'medium' },
+    (rootElement: HTMLElement) => expect(rootElement).toHaveClass('isRootSizeMedium'),
+  ],
+  [
+    { size: 'large' },
+    (rootElement: HTMLElement) => expect(rootElement).toHaveClass('isRootSizeLarge'),
+  ],
+
```
Comment (adamkudrna) [inline tests/propTests/sizePropTest.ts:14]: Needless blank line

File: tests/propTests_in_javacript/actionColorPropTest.ts
Comment (adamkudrna) [inline tests/propTests_in_javacript/actionColorPropTest.ts:1]: I suppose the `propTests_in_javascript` directory is not going to be merged into `master`, right?

File: .eslintrc [hunk 11]
```diff
@@ -1,11 +1,30 @@
 {
   "extends" : [
-    "@visionappscz/eslint-config-visionapps"
+    "@visionappscz/eslint-config-visionapps",
+    "airbnb-typescript",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:typescript-sort-keys/recommended",
+    // "plugin:@typescript-eslint/recommended-requiring-type-checking"
```
Comment (adamkudrna) [inline .eslintrc:7]: Is this line going to be used in a next step? If not, I'd delete it.

File: src/docs/_components/Placeholder/Placeholder.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        dark && styles.rootDark,
+        inline && styles.rootInline,
+      )}
+      style={style}
+    >
+      {children}
+    </div>
+  );
+};
+export default Placeholder;
```
Comment (adamkudrna) [inline src/docs/_components/Placeholder/Placeholder.tsx:34]: There should be a blank line before the default export, I believe.

File: src/lib/components/Alert/Alert.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        <span className={styles.closeSign}>×</span>
+      </button>
+      )}
+    </div>
+  );
+};
+
+export const AlertWithGlobalProps = withGlobalProps(Alert, 'Alert');
+
+export default AlertWithGlobalProps;
```
Comment (adamkudrna) [inline src/lib/components/Alert/Alert.tsx:62]: Please fix the missing end of line

**Review by adamkudrna — COMMENTED** [review]
I'm so curious about the final result, can't wait to finally learn TS 🙂.

- To me it appears **ESlint** is sleeping instead of doing its job… There are many formating issues _lost in translation_…
- **Please fix the branch name and commit message** according to [our conventions](https://react-ui.io/contribute/guidelines#git-workflow) so the changes make their way to the CHANGELOG (among other reasons 🙂).

File: definitions.d.ts (hunk 3 above)
Comment (mbohal) [reply definitions.d.ts]: > Does it make sense to separate them in TS?

Yes, it makes sense to do it now.

**Pull request closed without merging**